File: coding/Leetcode/DS_A/task_205.cpp
// https://leetcode.com/problems/isomorphic-strings/
// Created by ruslan515 on 22.06.23.
//

#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    bool isIso(string s, string t) {
        unordered_map<char, char> mp;
        bool answer = true;
        int n = s.size();
        for (int i = 0; i < n; ++i) {
            char ch1 = s[i];
            char ch2 = t[i];
            if (mp.find(ch2) != mp.end()) {
                char val = mp[ch2];
                if (val != ch1) {
                    answer = false;
                    break;
                }
            } else {
                mp[ch2] = ch1;
            }
        }
        return answer;
    }
    bool isIsomorphic(string s, string t) {
        bool answer = isIso(s, t) && isIso(t, s);
        return answer;
    }
};

int main() {
    bool answer;
    string s, t;
    Solution solve;

    s = "badc", t = "baba";
    answer = false;
    assert(answer == solve.isIsomorphic(s, t));

    s = "egg", t = "add";
    answer = true;
    assert(answer == solve.isIsomorphic(s, t));

    s = "foo", t = "bar";
    answer = false;
    assert(answer == solve.isIsomorphic(s, t));

    s = "paper", t = "title";
    answer = true;
    assert(answer == solve.isIsomorphic(s, t));

    return 0;
}

File: coding/Leetcode/Tasks/task_205.py
# https://leetcode.com/problems/ipo/
from typing import List, Optional


class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        answer = False
        n = len(s)
        m = len(t)
        if n != m:
            return answer
        answer = True
        s_to_t = {}
        t_to_s = {}
        for x, y in zip(s, t):
            # s_to_t
            if x in s_to_t.keys():
                val = s_to_t[x]
                if y != val:
                    answer = False
                    break
            else:
                s_to_t[x] = y

            if y in t_to_s.keys():
                val = t_to_s[y]
                if val != x:
                    answer = False
                    break
            else:
                    t_to_s[y] = x

        return answer


def main():
    solve = Solution()

    s = "badc"
    t = "baba"
    answer = False
    assert answer == solve.isIsomorphic(s, t)

    s = "egg"
    t = "add"
    answer = True
    assert answer == solve.isIsomorphic(s, t)

    s = "foo"
    t = "bar"
    answer = False
    assert answer == solve.isIsomorphic(s, t)

    s = "paper"
    t = "title"
    answer = True
    assert answer == solve.isIsomorphic(s, t)


if __name__ == "__main__":
    main()
